cluster/router/chain/chain.go

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package chain
+
+import (
+	"sort"
+	"sync"
+)
+
+import (
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/cluster/router"
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/protocol"
+)
+
+// RouterChain Router chain
+type RouterChain struct {
+	//full list of addresses from registry, classified by method name.
+	invokers []protocol.Invoker
+	//containing all routers, reconstruct every time 'route://' urls change.
+	routers []router.Router
+	// Fixed router instances: ConfigConditionRouter, TagRouter, e.g., the rule for each instance may change but the
+	// instance will never delete or recreate.
+	builtinRouters []router.Router
+
+	mutex sync.RWMutex
+}
+
+func (c RouterChain) Route(invoker []protocol.Invoker, url *common.URL, invocation protocol.Invocation) []protocol.Invoker {
+	finalInvokers := invoker
+	l := len(c.routers)
+	rs := make([]router.Router, l, l)
+	c.mutex.RLock()
+	copy(rs, c.routers)
+	c.mutex.RUnlock()
+
+	for _, r := range rs {
+		finalInvokers = r.Route(invoker, url, invocation)
```

Comment:
It seems that this statement should be `finalInvokers = r.Route(finalInvoker, url, invocation)`?